compiler/rustc_feature/src/builtin_attrs.rs

```diff
@@ -257,6 +257,7 @@ pub const BUILTIN_ATTRIBUTES: &[BuiltinAttribute] = &[
     ungated!(no_std, CrateLevel, template!(Word)),
     ungated!(no_implicit_prelude, Normal, template!(Word)),
     ungated!(non_exhaustive, AssumedUsed, template!(Word)),
+    // gated!(default, AssumedUsed, template!(Word), experimental!(derive_default_enum)),
```

Comment:
It was commented out because it caused a slew of errors regarding ambiguity. It obviously should not be eventually.

Ideally it would be a helper attribute as in `#[proc_macro_derive(Default, attributes(default))]`. How would I go about adding that plumbing? As far as I'm aware it doesn't currently exist for use in the compiler.

Reply:
Just trying some things again to remember exactly what happened.

```rust
gated!(Default, AssumedUsed, template!(Word), derive_default_enum, experimental!(derive_default_enum)),
```

When this line is used, the following error occurs:

```text
error[E0659]: `Default` is ambiguous (built-in attribute vs any other name)
   --> library/core/src/cmp.rs:585:38
    |
585 | #[derive(PartialEq, Eq, Debug, Copy, Default, Hash)]
    |                                      ^^^^^^^ ambiguous name
    |
    = note: `Default` could refer to a built-in attribute
```

Reply:
This line has a typo, `gated!(Default, ...)` -> `gated!(default, ...)`.
Declaring uppercase `Default` a built-in attribute will certainly cause ambiguities in practice.

Reply:
`sym::Default` actually is for the _lowercase_ `default` keyword [[ref](https://github.com/rust-lang/rust/blob/d2b04f075c0ce010758c4c8674152ff89d1d73f3/compiler/rustc_span/src/symbol.rs#L99)]. Introducing a new `sym::default` naturally conflicts.

I'd prefer to avoid this altogether if it's feasible (as you've indicated) to register a helper attribute of the `Default` derive macro. It would avoid this and inevitable errors with existing code.

Reply:
Worth noting that `kw::Default` is the lowercase symbol, which is what's desired here. I've no idea if that would work, as it would be contrary to literally every other built-in attribute.

Reply:
Helper attributes for built-in derive macros are implemented in https://github.com/rust-lang/rust/pull/87027.

Reply:
Awesome! Thanks for taking care of that bit of work.